pymatgen/transformations/advanced_transformations.py

```diff
@@ -1360,3 +1361,213 @@ def inverse(self):
+
+            # minimum distance any direction of the supercell must be as large as
+            hard_sc_size_threshold = self.nn_dist * self.num_nn_cutoff
+
+            # target_threshold is used as the desired cubic side lengths of the supercell
+            target_sc_size = hard_sc_size_threshold
+
+            while sc_not_found:
+                target_sc_lat_vecs = np.eye(3, 3) * target_sc_size
+
+                trans_mat = np.linalg.inv(self.lat_vecs) @ target_sc_lat_vecs
+
+                # round the entries of T and force T to be nonsingular
+                trans_mat = self._round_and_make_arr_singular(trans_mat)
+
+                proposed_sc_lat_vecs = trans_mat @ self.lat_vecs
+
+                # Check how many nearest neighbors the proposed supercell allows
+                scBasesNorms = [np.linalg.norm(proposed_sc_lat_vecs[0]),
+                                np.linalg.norm(proposed_sc_lat_vecs[1]),
+                                np.linalg.norm(proposed_sc_lat_vecs[2])]
+                maxNorm = max(scBasesNorms)
+                maxIndex = scBasesNorms.index(maxNorm)
+                idx = list(range(3))
+                idx.remove(maxIndex)
+                a = proposed_sc_lat_vecs[maxIndex]
+                b = proposed_sc_lat_vecs[idx[0]]
+                c = proposed_sc_lat_vecs[idx[1]]
+                projb_a = self.proj(b, a)
+                projc_a = self.proj(c, a)
+
+                if np.linalg.norm(projb_a) > np.linalg.norm(projc_a):
+                    length = np.linalg.norm(a - projb_a)
+                else:
+                    length = np.linalg.norm(a - projc_a)
+                width = math.sqrt(np.linalg.norm(b) ** 2 - np.linalg.norm(projb_a) ** 2)
+                ab_normal = np.cross(a, b)  # get normal direction from AB plane
+                height = np.linalg.norm(self.proj(c, ab_normal))  # project c onto AB plane normal
+
+                self.smallest_dim = min([length, width, height])
+
+                # Get number of atoms
+                superstructure = SupercellTransformation(trans_mat).apply_transformation(self.structure)
+                num_at = superstructure.num_sites
+
+                if self.smallest_dim >= hard_sc_size_threshold \
+                        and num_at >= self.min_atoms and num_at <= self.max_atoms:
+                    return superstructure, trans_mat
+                else:
+                    # Increase threshold until proposed supercell meets requirements
+                    target_sc_size += 0.1
+                    if num_at > self.max_atoms:
+                        raise AttributeError('While trying to solve for the supercell, the max'
+                                             'number of atoms was exceeded. Try lowering the number'
+                                             'of nearest neighbor distances.')
+
+class PerturbedSupercellsTransformation:
```

Comment:
Also, see my notes for the cubic transformation:

This class doesn't subclass `AbstractTransformation`.

E.g. see `class SupercellTransformation(AbstractTransformation):` Note the `(AbstractTransformation)`.

Furthermore, it doesn't implement the required functions that all abstract transformations should. See `SupercellTransformation` for an example. E.g., you should add and modify the following:

```python
    def apply_transformation(self, structure):
        return structure * self.scaling_matrix

    def __str__(self):
        return "Supercell Transformation with scaling matrix " + \
            "{}".format(self.scaling_matrix)

    def __repr__(self):
        return self.__str__()

    @property
    def inverse(self):
        raise NotImplementedError()

    @property
    def is_one_to_many(self):
        return False
```

Read the `AbstractTransformation` docs for more information.